rllib/core/rl_trainer/trainer_runner.py

```diff
@@ -128,20 +135,52 @@ def _distributed_update(self, batch: MultiAgentBatch) -> List[Mapping[str, Any]]
         Returns:
             A list of dictionaries of results from the updates from the RLTrainer(s)
         """
-        refs = []
+
+        if block:
+            batches = self._shard_sample_batch(batch)
+            results = self._worker_manager.foreach_actor(
+                [lambda w: w.update(b) for b in batches]
+            )
+        else:
+            if batch is not None:
+                self._in_queue.append(batch)
+            results = self._worker_manager.fetch_ready_async_reqs()
+            if self._worker_manager_ready() and self._in_queue:
+                batch = self._in_queue.popleft()
+                batches = self._shard_sample_batch(batch)
+                self._worker_manager.foreach_actor_async(
+                    [lambda w: w.update(b) for b in batches]
+                )
+
+        return self._get_results(results)
+
+    def _worker_manager_ready(self):
+        return self._worker_manager.num_outstanding_async_reqs() == 0
+
+    def _shard_sample_batch(self, batch: MultiAgentBatch) -> List[MultiAgentBatch]:
```

Comment:
Let's convert this to an iterator and put it [here](https://github.com/ray-project/ray/blob/master/rllib/utils/minibatch_utils.py), similar to others? We need to consolidate the iterator hierarchy tho, which should be an easy thing to do.

Reply:
done